clang/test/SemaSYCL/unique_stable_name.cpp

```diff
@@ -0,0 +1,119 @@
+// RUN: %clang_cc1 %s -fsyntax-only -ast-dump -fsycl-is-device -triple spir64 -verify
+
+// expected-error 4 @+2{{kernel instantiation changes the result of an evaluated __builtin_unique_stable_name}}
+template <typename KernelName, typename KernelType>
+[[clang::sycl_kernel]] void kernel_single_task(KernelType kernelFunc) {
+  kernelFunc();
+}
+template <template <typename> typename Outer, typename Inner>
+struct S {
+  void operator()() const;
+};
+
+template <typename Ty>
+struct T {};
+
+template <typename Func>
+void kernel3func(const Func &F3) {
+  // expected-note@+1{{__builtin_unique_stable_name evaluated here}}
+  constexpr const char *F3_output = __builtin_unique_stable_name(F3);
+  // expected-error@+1{{error: kernel instantiation changes the result of an evaluated __builtin_unique_stable_name}}
+  kernel_single_task<class kernel3>(F3);
+}
+
+template <typename Func>
+void kernel4func(const Func &F4) {
+  // expected-note@+1{{__builtin_unique_stable_name evaluated here}}
+  constexpr const char *F4_output = __builtin_unique_stable_name(F4);
+  // expected-error@+1{{error: kernel instantiation changes the result of an evaluated __builtin_unique_stable_name}}
+  kernel_single_task<class kernel4>([]() {});
+}
+
+template <typename Ty>
+auto func() -> decltype(__builtin_unique_stable_name(Ty::str));
+
+struct Derp {
+  static constexpr const char str[] = "derp derp derp";
+};
+
+template <typename T>
+static constexpr const char *output1 = __builtin_unique_stable_name(T);
+
+#define MACRO12()                \
+  auto l12 = []() { return 1; }; \
+  constexpr const char *l1_output = __builtin_unique_stable_name(l12);
+
+int main() {
+  // kernel0
+  kernel_single_task<class kernel0>( // func<Derp>());
+      []() {
+        func<Derp>();
```

Comment:
Didn't we want this to be an error because it means that `__builtin_unique_stable_name` will appear in a mangling (due to being in a trailing return type)?

Reply:
I see!  Currently that asserts with an unreachable.  I'll put together a patch to fix this, but we should definitely check in a codegen test what the mangling of something like that looks like.

Reply:
Added CodeGen test for the mangling